### PR TITLE
Proposal: YurtAppDaemon

### PR DESCRIPTION
**Pull request by kadisi**
Signed-off-by: bingyu.zj <bingyu.zj@alibaba-inc.com>

 /kind feature

#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note

```

### CODE REVIEW

**Comment by rambohe-ch** [comment]
@kadisi Based on the proposal, An idea comes out that combine `UnitedDeployment` and `UnitedDaemonSet`, and we can add some fields in `UnitedDeployment` specify that create/delete Deployments/Statefulset along with create/delete nodePool.

**Review by zzguang — COMMENTED** [review]
@kadisi , as wenjun mentioned about the united-service feature request on last community meeting, the uniteddaemonset workload template will add ServiceTemplate support in future, right?

**Comment by gnunu** [comment]
> @kadisi Based on the proposal, An idea comes out that combine `UnitedDeployment` and `UnitedDaemonSet`, and we can add some fields in `UnitedDeployment` specify that create/delete Deployments/Statefulset along with create/delete nodePool.

for daemonset, another question. now UnitedDeployment supports StatefulSet and Deployment. if DaemonSet (one pod per node in nodepool) should be supported also? i think it has its scenarios.

if the above thought is reasonable, then there is something subtle about naming here: daemonset means one deployment per nodepool (pool level daemonset), or one pod per node (node level daemonset)? 

sounds UnitedDaemonSet can have some stress on nodepool level, and thus are at the same level as UnitedDeployment.

**Comment by Fei-Guo** [comment]
I can see the difference between UnitedDeamonset and UnitedDeployment as the former can automatically deploy workload when  nodepool is created. The benefit is that we can could avoid manually changing UnitedDeployment had it been used for the same purpose. The naming maybe misleading, how about NodePoolDaemonSet?

**Comment by kadisi** [comment]
> @kadisi Based on the proposal, An idea comes out that combine `UnitedDeployment` and `UnitedDaemonSet`, and we can add some fields in `UnitedDeployment` specify that create/delete Deployments/Statefulset along with create/delete nodePool.

@rambohe-ch This is not necessary, UnitedDeployment is not strongly associated with NodePool。If you delete a Pool in UnitedDeployment,sub Deployments/Statefulset will be automatically deleted

**Comment by kadisi** [comment]
> for daemonset, another question. now UnitedDeployment supports StatefulSet and Deployment. if DaemonSet (one pod per node in nodepool) should be supported also? i think it has its scenarios.
> 
> if the above thought is reasonable, then there is something subtle about naming here: daemonset means one deployment per nodepool (pool level daemonset), or one pod per node (node level daemonset)?
> 
> sounds UnitedDaemonSet can have some stress on nodepool level, and thus are at the same level as UnitedDeployment.

@gnunu 
for Daemonset, there is no need for more encapsulation; it can be done on its own.

**Comment by kadisi** [comment]
> @kadisi , as wenjun mentioned about the united-service feature request on last community meeting, the uniteddaemonset workload template will add ServiceTemplate support in future, right?

@zzguang yes

File: docs/enhancements/20210729-yurtappdaemon.md [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+### Goals
+- Define the API of YurtAppDaemon 
+- Provide YurtAppDaemon controller
+- Provide YurtAppDaemon webhook 
+
+## Proposal
+
+- What is the plan for implementing this feature?
+
+  Provide yurtunit-manager operator to YurtAppDaemon CRD
```
Comment (rambohe-ch) [inline docs/enhancements/20210729-yurtappdaemon.md:44]: yurtunit-manager --> yurt-app-manager?

File: docs/enhancements/20210729-yurtappdaemon.md [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+	Spec              appsv1.DeploymentSpec `json:"spec"`
+}
+
+
+```
+
+## Implementation History
+
++ [ ] : YurtAppDaemon api crd
++ [ ] : yurtunit-manager YurtAppDaemon controller
```
Comment (rambohe-ch) [inline docs/enhancements/20210729-yurtappdaemon.md:189]: yurtunit-manager --> yurt-app-manager?

File: docs/enhancements/20210729-yurtappdaemon.md [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+
+// YurtAppDaemonConditionType indicates valid conditions type of a YurtAppDaemon.
+type YurtAppDaemonConditionType string
+
+const (
+	// WorkLoadProvisioned means all the expected workload are provisioned
+	WorkLoadProvisioned YurtAppDaemonConditionType = "WorkLoadProvisioned"
+	// WorkLoadUpdated means all the workload are updated.
+	WorkLoadUpdated YurtAppDaemonConditionType = "WorkLoadUpdated"
+	// WorkLoadFailure is added to a UnitedDeployment when one of its workload has failure during its own reconciling.
```
Comment (rambohe-ch) [inline docs/enhancements/20210729-yurtappdaemon.md:63]: UnitedDeployment --> YurtAppManager?

File: docs/enhancements/20210729-yurtappdaemon.md [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+}
+
+// +genclient
+// +kubebuilder:object:root=true
+// +kubebuilder:subresource:status
+// +kubebuilder:resource:shortName=udd
+// +kubebuilder:printcolumn:name="WorkloadTemplate",type="string",JSONPath=".status.templateType",description="The WorkloadTemplate Type."
+// +kubebuilder:printcolumn:name="AGE",type="date",JSONPath=".metadata.creationTimestamp",description="CreationTimestamp is a timestamp representing the server time when this object was created. It is not guaranteed to be set in happens-before order across separate operations. Clients may not set this value. It is represented in RFC3339 form and is in UTC."
+
+// YurtAppDaemon is the Schema for the uniteddeployments API
```
Comment (rambohe-ch) [inline docs/enhancements/20210729-yurtappdaemon.md:139]: uniteddeployment --> yurtappdaemons?

File: docs/enhancements/20210729-yurtappdaemon.md [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+// +kubebuilder:object:root=true
+
+// YurtAppDaemonList contains a list of YurtAppDaemon
+type YurtAppDaemonList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+	Items           []YurtAppDaemon `json:"items"`
+}
+
+// WorkloadTemplate defines the pool template under the UnitedDeployment.
```
Comment (rambohe-ch) [inline docs/enhancements/20210729-yurtappdaemon.md:157]: UnitedDeployment --> YurtAppDaemon?

**Comment by DrmagicE** [comment]
How about adding patch feature to YurtAppDaemon, just like we did in YurtAppSet.

**Comment by rambohe-ch** [comment]
> How about adding patch feature to YurtAppDaemon, just like we did in YurtAppSet.

@DrmagicE I think we'd better implement general YurtAppDaemon at first. and after that we can discuss  whether add patch feature or not based on requirements from end users.

**Comment by DrmagicE** [comment]
@rambohe-ch OK, It is reasonable.

**Pull request merged**